File: apps/backend/src/api/schemas/youtube.py
"""Pydantic schemas for YouTube import API."""
from typing import Optional

from pydantic import BaseModel


class AvailableSubtitle(BaseModel):
    """Available subtitle track for a video."""

    lang_code: str
    label: str
    is_auto: bool


class YouTubePreviewResponse(BaseModel):
    """Response for GET /youtube/preview endpoint."""

    video_id: str
    title: str
    duration_ms: Optional[int]
    channel_name: Optional[str]
    thumbnail_url: Optional[str]
    available_subtitles: list[AvailableSubtitle]


class YouTubeImportRequest(BaseModel):
    """Request body for POST /youtube/import endpoint."""

    url: str
    title: str
    language_id: int
    subtitle_lang_code: str
    use_auto_captions: bool


class YouTubeImportResponse(BaseModel):
    """Response for POST /youtube/import endpoint (202 Accepted)."""

    video_id: str
    content_item_id: str
    status: str


class YouTubeSubtitleUploadResponse(BaseModel):
    """Response for POST /youtube/{video_id}/subtitles/upload endpoint."""

    video_id: str
    status: str
    lines_parsed: int


class YouTubeStatusResponse(BaseModel):
    """Response for GET /youtube/{video_id}/status endpoint."""

    video_id: str
    content_item_id: str
    status: str
    error_message: Optional[str] = None

File: apps/web/src/lib/api/youtube.ts
import { apiClient } from "./client"

export interface AvailableSubtitle {
  lang_code: string
  label: string
  is_auto: boolean
}

export interface YouTubePreview {
  video_id: string
  title: string
  duration_ms: number | null
  channel_name: string | null
  thumbnail_url: string | null
  available_subtitles: AvailableSubtitle[]
}

export interface YouTubeImportResponse {
  video_id: string
  content_item_id: string
  status: string
}

export async function previewYouTube(url: string): Promise<YouTubePreview> {
  return apiClient(`/youtube/preview?url=${encodeURIComponent(url)}`)
}

export async function importYouTube(opts: {
  url: string
  title: string
  language_id: number
  subtitle_lang_code: string
  use_auto_captions: boolean
}): Promise<YouTubeImportResponse> {
  return apiClient("/youtube/import", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(opts),
  })
}
